File: q2_ms/xcms/adjust_retention_time_obiwarp.py
import copy

from q2_ms.types import XCMSExperimentDirFmt, mzMLDirFmt
from q2_ms.utils import run_r_script
from q2_ms.xcms.utils import change_data_paths


def adjust_retention_time_obiwarp(
    spectra: mzMLDirFmt,
    xcms_experiment: XCMSExperimentDirFmt,
    bin_size: float = 25,
    center_sample: str = None,
    response: float = 1,
    dist_fun: str = "cor_opt",
    gap_init: float = None,
    gap_extend: float = None,
    factor_diag: float = 2,
    factor_gap: float = 1,
    local_alignment: bool = False,
    init_penalty: float = None,
    sample_metadata_column: str = None,
    subset_label: str = None,
    subset_adjust: str = None,
    rtime_difference_threshold: float = 5,
    chunk_size: int = 1,
    threads: int = 1,
) -> XCMSExperimentDirFmt:
    _validate_parameters(
        sample_metadata_column,
        subset_label,
        subset_adjust,
        local_alignment,
        init_penalty,
    )
    # Create parameters dict
    params = copy.copy(locals())

    # Innit XCMSExperimentDirFmt
    xcms_experiment_aligned = XCMSExperimentDirFmt()

    # Change data paths in xcms experiment
    change_data_paths(str(xcms_experiment), str(spectra))

    # Add output path to params
    params["output_path"] = str(xcms_experiment_aligned)

    # Run R script
    run_r_script(params, "adjust_retention_time_obiwarp", "XCMS")

    return xcms_experiment_aligned


def _validate_parameters(
    sample_metadata_column, subset_label, subset_adjust, local_alignment, init_penalty
):
    if bool(sample_metadata_column) != bool(subset_label) or bool(
        sample_metadata_column
    ) != bool(subset_adjust):
        raise ValueError(
            'The parameters "--p-sample-metadata-column",  "--p-subset-label" and '
            '"--p-subset-adjust" have to be used in combination with each other.'
        )

    if not local_alignment and init_penalty:
        raise ValueError(
            'The parameter "--p-init-penalty" can only be used if '
            '"--p-local-alignment" is set to True.'
        )

File: q2_ms/assets/adjust_retention_time_obiwarp.R
#!/usr/local/bin/Rscript --vanilla

library(xcms)
library(MsIO)
library(Spectra)
library(optparse)
library(MsExperiment)

# Define command-line options
option_list <- list(
  make_option(opt_str = "--spectra", type = "character"),
  make_option(opt_str = "--xcms_experiment", type = "character"),
  make_option(opt_str = "--bin_size", type = "numeric"),
  make_option(opt_str = "--center_sample", type = "character"),
  make_option(opt_str = "--response", type = "numeric"),
  make_option(opt_str = "--dist_fun", type = "character"),
  make_option(opt_str = "--gap_init", type = "numeric"),
  make_option(opt_str = "--gap_extend", type = "numeric"),
  make_option(opt_str = "--factor_diag", type = "numeric"),
  make_option(opt_str = "--factor_gap", type = "numeric"),
  make_option(opt_str = "--local_alignment", type = "logical"),
  make_option(opt_str = "--init_penalty", type = "numeric"),
  make_option(opt_str = "--sample_metadata_column", type = "character"),
  make_option(opt_str = "--subset_label", type = "character"),
  make_option(opt_str = "--subset_adjust", type = "character"),
  make_option(opt_str = "--rtime_difference_threshold", type = "numeric"),
  make_option(opt_str = "--threads", type = "integer"),
  make_option(opt_str = "--output_path", type = "character"),
  make_option(opt_str = "--chunk_size", type = "integer")
)

# Parse arguments
optParser <- OptionParser(option_list = option_list)
opt <- parse_args(optParser)

# Import XcmsExperiment or MsExperiment depending on previous data processing steps
XCMSExperiment <- tryCatch({
    readMsObject(XcmsExperiment(), PlainTextParam(opt$xcms_experiment))
}, error = function(e) {
    readMsObject(MsExperiment(), PlainTextParam(opt$xcms_experiment))
})

# Set parameters
ObiwarpParams <- ObiwarpParam(
  binSize = opt$bin_size,
  response = opt$response,
  distFun = opt$dist_fun,
  factorDiag = opt$factor_diag,
  factorGap = opt$factor_gap,
  localAlignment = opt$local_alignment,
  initPenalty = opt$init_penalty,
  rtimeDifferenceThreshold = opt$rtime_difference_threshold
)

if (!is.null(opt$sample_metadata_column)) {
    ObiwarpParams@subset <- which(sampleData(XCMSExperiment)[[opt$sample_metadata_column]] == opt$subset_label)
}
if (!is.null(opt$center_sample)) {
    ObiwarpParams@centerSample <- which(sampleData(XCMSExperiment)[[1]]== opt$center_sample)
}
if (!is.null(opt$gap_init)) ObiwarpParams@gapInit <- opt$gap_init
if (!is.null(opt$gap_extend)) ObiwarpParams@gapExtend <- opt$gap_extend
if (!is.null(opt$subset_adjust)) ObiwarpParams@subsetAdjust <- opt$subset_adjust

# Adjust retention time using the Obiwarp algorithm
XCMSExperimentRT <- adjustRtime(
  object = XCMSExperiment,
  param = ObiwarpParams,
  chunkSize = opt$chunk_size,
  BPPARAM = MulticoreParam(workers = opt$threads)
)

# Export the XCMSExperiment object to the directory format
saveMsObject(XCMSExperimentRT, param = PlainTextParam(path = opt$output_path))
